### ISSUE TITLE
Reproducible ImageNet

### ISSUE THREAD

**Issue by mohitreddy1996** [google/flax#291]
Currently ImageNet example (one of others) is not reproducible. This is mainly due to the stochastic nature of the following:
- Multi host (and per host, multi threaded) training.
- Shuffle operations in input pipeline (shuffle files, shuffle examples before batching etc).
- Stateful seed in TF OPs like `tf.image.sample_distorted_bounding_box`
- Model parameters initialization.

Imagenet example uses a global random seed (statically set to 0) used to create a JAX PRNGKey.

Few of the issues are solvable using `jax.random.split` and performing stateless folding-in the RNG keys with host ids. Although doing something similar for TF OPs is problematic due to stateful randomness nature of OPs and lack of stable APIs for creating/performing stateless splits of RNG keys (few stateless APIs exist in `tf.random.stateless_`, although `stateless_split` and `stateless_fold_in` are available as experimental APIs in nightly builds).

**Comment by Marvin182** [comment]
This is doable but currently involves quite a bit of extra code. This will become easier in the future. We can keep the issue.

**Comment by marcvanzee** [comment]
It is really important to us (and JAX) to have fully reproducible experiments. In fact, this is one of the motivations of JAX's RNG design. However, it seems better to write a HOWTO on how to set up a fully reproducible experiment, rather than forcing all our examples to be fully reproducible. I have created a new issue https://github.com/google/flax/issues/1851 for this and I'm closing this one.
